File: Searching/JavaCodes/Leetcode875.java
class Solution {
   public static boolean canIEatWithKCapacity(int[] nums,int capacity,int hours){
        for(int i=0;i< nums.length;i++){
            if(hours<0){
                return false;
            }
            int bananas=nums[i];
            int hourToFinish=(bananas+capacity-1)/capacity;
            hours-=hourToFinish;
        }
        return hours>=0;
    }

    public static int minEatingSpeed(int[] piles, int h) {
        int max=0;

        for (int i = 0; i < piles.length; i++) {
            max=Math.max(max,piles[i]);
        }
        int start=1,end=max;

        while(start<end){
            int capacity=start+(end-start)/2;
            if(canIEatWithKCapacity(piles,capacity,h)){
                end=capacity;
            }
            else{
                start=capacity+1;
            }
        }
        return start;
    }
}

File: LogicOverSyntaxCodes/Searching/PythonCodes/Leetcode875.py
class Solution(object):
    def canIEatWithKCapacity(self, piles, capacity, hours):
        for bananas in piles:
            if hours < 0:
                return False
            hour_to_finish = (bananas + capacity - 1) // capacity
            hours -= hour_to_finish
        return hours >= 0

    def minEatingSpeed(self, piles, h):
        max_capacity = max(piles)
        start, end = 1, max_capacity

        while start < end:
            capacity = start + (end - start) // 2
            if self.canIEatWithKCapacity(piles, capacity, h):
                end = capacity
            else:
                start = capacity + 1

        return start
